[RISCV] Reverse (add x, (zext c)) back to (select c, (add x, 1), x)

lukel97 opened the pull request:
A common induction variable pattern that seems to be emitted by the loop vectorizer is a vadd.vv where one of the operands is a vmerge.vim of zeroes and ones:

```asm
          vmv.v.i v12, 0
  .LBB9_109:                              # %vector.body
          vsetvli zero, zero, e8, mf4, ta, ma
          vluxei64.v      v9, (t1), v10
          vmseq.vi        v0, v9, 0
          vsetvli zero, zero, e32, m1, ta, ma
          vmerge.vim      v9, v12, 1, v0
          vadd.vv v8, v8, v9
```

I'm not sure if this is what it generates directly, but in any case InstCombine will transform `select c, (add x, 1), x` to `add x, (zext c)`.

On RISC-V though we don't have a native instruction for zero extending i1 elements, and it gets lowered as a vmerge.vim and vmv.v.i instead.

We can reverse this transform so that we pull the select outside of the binary op, which allows us to fold it into a masked op:

```asm
      vadd.vi v8, v8, 1, v0.t
```

Specifically, we can do this transform for any binary op where the identity is zero.

This pattern doesn't show up in the in-tree tests, but shows up more frequently in llvm-test-suite/SPEC CPU 2017:
[llvm-test-suite.diff.txt](https://github.com/llvm/llvm-project/files/14823564/zext-binop.diff.txt)

Alive2 proof: https://alive2.llvm.org/ce/z/VKFegj

lukel97 committed:
[RISCV] Reverse (add x, (zext c)) back to (select c, (add x, 1), x)

A common induction variable pattern that seems to be emitted by the loop vectorizer is a vadd.vv where one of the operands is a vmerge.vim of zeroes and ones:

	  vmv.v.i v12, 0
  .LBB9_109:                              # %vector.body
	  vsetvli zero, zero, e8, mf4, ta, ma
	  vluxei64.v      v9, (t1), v10
	  vmseq.vi        v0, v9, 0
	  vsetvli zero, zero, e32, m1, ta, ma
	  vmerge.vim      v9, v12, 1, v0
	  vadd.vv v8, v8, v9

I'm not sure if this is what it generates directly, but in any case InstCombine will transform `select c, (add x, 1), x` to `add x, (zext c)`.

On RISC-V though we don't have a native instruction for zero extending i1 elements, and it gets lowered as a vmerge.vim and vmv.v.i instead.

We can reverse this transform so that we pull the select outside of the binary op, which allows us to fold it into a masked op:

      vadd.vi v8, v8, 1, v0.t

Specifically, we can do this transform for any binary op where the identity is zero.

Alive2 proof: https://alive2.llvm.org/ce/z/VKFegj

lukel97 committed:
Add tests for converting i1 zext into select in RISCVCodeGenPrepare

dtcxzyw commented on llvm/lib/Target/RISCV/RISCVCodeGenPrepare.cpp:
```diff
@@ -52,12 +52,59 @@ class RISCVCodeGenPrepare : public FunctionPass,
+  IRBuilder<> Builder(&BO);
+  Value *Splat = Builder.CreateVectorSplat(
+      Ty->getElementCount(), ConstantInt::get(Ty->getElementType(), 1));
+  Value *Add = Builder.CreateAdd(RHS, Splat);
```
Can we also handle `select c, (or/xor x, 1), x -> or/xor x, (zext c)`?

dtcxzyw commented on llvm/lib/Target/RISCV/RISCVCodeGenPrepare.cpp:
```diff
@@ -52,12 +52,59 @@ class RISCVCodeGenPrepare : public FunctionPass,
+///
+///       vadd.vi v8, v8, 1, v0.t
+bool RISCVCodeGenPrepare::visitAdd(BinaryOperator &BO) {
+  VectorType *Ty = dyn_cast<VectorType>(BO.getType());
+  if (!Ty)
+    return false;
```
```suggestion
  if (!BO.getType()->isVectorTy())
    return false;
```

BTW I guess you don't want to handle fp binops :)

lukel97 replied:
Isn't a zext always on integer types?

dtcxzyw replied:
I mean you can take an early exit here :)

dtcxzyw commented on llvm/lib/Target/RISCV/RISCVCodeGenPrepare.cpp:
```diff
@@ -52,12 +52,59 @@ class RISCVCodeGenPrepare : public FunctionPass,
+  if (!match(&BO, m_c_BinOp(m_OneUse(m_ZExt(m_Value(Mask))), m_Value(RHS))))
+    return false;
+
+  if (!cast<VectorType>(Mask->getType())->getElementType()->isIntegerTy(1))
```
```suggestion
  if (!Mask->getType()->isIntOrIntVectorTy(1))
```

We have checked that `Mask` is a vector.

dtcxzyw commented on llvm/lib/Target/RISCV/RISCVCodeGenPrepare.cpp:
```diff
@@ -52,12 +52,59 @@ class RISCVCodeGenPrepare : public FunctionPass,
+    return false;
+
+  IRBuilder<> Builder(&BO);
+  Value *Splat = Builder.CreateVectorSplat(
+      Ty->getElementCount(), ConstantInt::get(Ty->getElementType(), 1));
```
```suggestion
  Value *Splat = ConstantInt::get(BO.getType(), 1);
```

dtcxzyw commented on llvm/lib/Target/RISCV/RISCVCodeGenPrepare.cpp:
```diff
@@ -52,12 +52,59 @@ class RISCVCodeGenPrepare : public FunctionPass,
+/// InstCombine will canonicalize selects of binary ops where the identity is
+/// zero to zexts:
+///
+/// select c, (add x, 1), x -> add x, (zext c)
```
Can we also handle `select c, (sub x, 1), x -> add x, (sext c)`?

dtcxzyw commented on llvm/lib/Target/RISCV/RISCVCodeGenPrepare.cpp:
```diff
@@ -52,12 +52,59 @@ class RISCVCodeGenPrepare : public FunctionPass,
+  using namespace PatternMatch;
+
+  Value *Mask, *RHS;
+  if (!match(&BO, m_c_BinOp(m_OneUse(m_ZExt(m_Value(Mask))), m_Value(RHS))))
```
Please add a test that commutes operands and a negative test with multi-use zext.

dtcxzyw commented on llvm/lib/Target/RISCV/RISCVCodeGenPrepare.cpp:
```diff
@@ -52,12 +52,59 @@ class RISCVCodeGenPrepare : public FunctionPass,
+  if (!Ty)
+    return false;
+
+  Constant *Identity = ConstantExpr::getIdentity(&BO, BO.getType());
+  if (!Identity->isZeroValue())
+    return false;
```
```suggestion
```
The identity is always zero.

lukel97 replied:
Argh, I meant to make this generalised to all bin ops. In which case we need to check the identity, since mul and and etc. don't have zero identity.

lukel97 replied:
Should be generalized across bin ops now. Whoops

lukel97 committed:
Generalize from add to bin op (whoops)
Use isVectorTy/isIntOrIntVectorTy
Add commutated, multi-use and or/sub test cases

lukel97 commented on llvm/lib/Target/RISCV/RISCVCodeGenPrepare.cpp:
```diff
@@ -52,12 +52,58 @@ class RISCVCodeGenPrepare : public FunctionPass,
+    return false;
+
+  // TODO: We could allow sub if we did a non-commutative match
+  Constant *Identity = ConstantExpr::getIdentity(&BO, BO.getType());
```
I've left the sub case for later, since I believe InstCombine canonicalizes `sub x, (zext c)` to `add x, (sext c)` anyway. Definitely something that the loop vectorizer could be emitting though

dtcxzyw commented on llvm/lib/Target/RISCV/RISCVCodeGenPrepare.cpp:
```diff
@@ -52,12 +52,58 @@ class RISCVCodeGenPrepare : public FunctionPass,
+
+  // TODO: We could allow sub if we did a non-commutative match
+  Constant *Identity = ConstantExpr::getIdentity(&BO, BO.getType());
+  if (!Identity || !Identity->isZeroValue())
```
```suggestion
  if (!Identity || !Identity->isNullValue())
```

dtcxzyw commented on llvm/lib/Target/RISCV/RISCVCodeGenPrepare.cpp:
```diff
@@ -52,12 +52,58 @@ class RISCVCodeGenPrepare : public FunctionPass,
+
+  IRBuilder<> Builder(&BO);
+  Value *Splat = ConstantInt::get(BO.getType(), 1);
+  Value *BinOp = Builder.CreateBinOp(BO.getOpcode(), RHS, Splat);
```
Should we copy poison-generating flags?

lukel97 replied:
Good point, let me check on alive. We may not need to because the select stops poison from propagating

lukel97 replied:
Doesn't look like it's needed for correctness, but probably better to copy them to retain the information: https://alive2.llvm.org/ce/z/9PvEdE

lukel97 committed:
Copy IR flags
Use isNullValue

topperc commented on llvm/lib/Target/RISCV/RISCVCodeGenPrepare.cpp:
```diff
@@ -52,12 +52,60 @@ class RISCVCodeGenPrepare : public FunctionPass,
 
 } // end anonymous namespace
 
+/// InstCombine will canonicalize selects of binary ops where the identity is
```
Why does this need to be done on IR instead of SelectionDAG?

lukel97 replied:
Just seemed like the right place to put it since it's undoing an InstCombine transform. I don't think there should be any issue doing it as a DAG combine. Would that be preferred?

topperc commented on llvm/test/CodeGen/RISCV/rvv/fixed-vectors-vaaddu.ll:
```diff
@@ -168,13 +168,14 @@ define <8 x i64> @vaaddu_vv_v8i64_floor(<8 x i64> %x, <8 x i64> %y) {
-; CHECK-NEXT:    vsetivli zero, 8, e8, mf2, ta, ma
-; CHECK-NEXT:    vmv.v.i v9, 0
-; CHECK-NEXT:    vmerge.vim v10, v9, 1, v0
+; CHECK-NEXT:    vmv1r.v v9, v0
```
Is this an improvement?

topperc commented on llvm/lib/Target/RISCV/RISCVCodeGenPrepare.cpp:
```diff
@@ -52,12 +52,60 @@ class RISCVCodeGenPrepare : public FunctionPass,
+
+  IRBuilder<> Builder(&BO);
+  Value *Splat = ConstantInt::get(BO.getType(), 1);
+  Value *NewBO = Builder.CreateBinOp(BO.getOpcode(), RHS, Splat);
```
This requires a freeze on RHS

lukel97 replied:
I think I'm missing something here. We aren't increasing the number of uses of RHS and alive doesn't seem to complain. And we're not speculatively executing anything?

topperc replied:
Isn't `RHS` used by `NewBO` and `Select`?

lukel97 replied:
Sorry yes, I missed that. 

I'm guessing the reason why alive doesn't require the freeze then is because `(add (zext c), undef) ~> undef`. So `select c, (add undef, 1), undef` would also be just as undefined?

dtcxzyw replied:
Freeze is unnecessary here because we only use one arm of the select inst.

topperc replied:
Why does `foldBinOpIntoSelect` use a freeze then?

dtcxzyw replied:
> Why does `foldBinOpIntoSelect` use a freeze then?

https://github.com/llvm/llvm-project/blob/8bb9443333e0117ab61feecce9de339b11b924fc/llvm/lib/CodeGen/SelectionDAG/DAGCombiner.cpp#L2375-L2388

IIUC the freeze can be eliminated as well.

dtcxzyw replied:
Related patch: https://reviews.llvm.org/D118644

topperc commented:
I think we already have most of a better version in this transform in SelectionDAG

Once we expand to (add (vselect (splat_vector 1), (splat_vector 0))), `foldBinOpIntoSelect` in DAGCombiner is almost able to see the (splat_vector 0) as being the identity constant for the add. The only reason it doesn't is because `isNeutralConstant` doesn't pass `AllowTruncation=true` to `isConstOrConstSplat`.

As a quick hack I tried

```
diff --git a/llvm/lib/CodeGen/SelectionDAG/SelectionDAG.cpp b/llvm/lib/CodeGen/SelectionDAG/SelectionDAG.cpp
index e8d1ac1d3a91..9ff7a2d7ddf4 100644
--- a/llvm/lib/CodeGen/SelectionDAG/SelectionDAG.cpp
+++ b/llvm/lib/CodeGen/SelectionDAG/SelectionDAG.cpp
@@ -11549,7 +11549,7 @@ bool llvm::isNeutralConstant(unsigned Opcode, SDNodeFlags Flags, SDValue V,
                              unsigned OperandNo) {
   // NOTE: The cases should match with IR's ConstantExpr::getBinOpIdentity().
   // TODO: Target-specific opcodes could be added.
-  if (auto *Const = isConstOrConstSplat(V)) {
+  if (auto *Const = isConstOrConstSplat(V, false, /*AllowTruncation*/Opcode == ISD::ADD)) {
     switch (Opcode) {
     case ISD::ADD:
     case ISD::OR:
```

lukel97 commented:
Thanks for pointing that out, seems better to try and fix foldSelectWithIdentityConstant. Will take a look
